cmd/encryption-v1.go

```diff
@@ -123,8 +123,6 @@ func ParseSSECustomerHeader(header http.Header) (key []byte, err error) {
 
 // This function rotates old to new key.
 func rotateKey(oldKey []byte, newKey []byte, bucket, object string, metadata map[string]string) error {
-	delete(metadata, crypto.SSECKey) // make sure we do not save the key by accident
```

Comment:
why are you removing this?

Reply:
> why are you removing this?

The rule is not clear when to remove and when not, so the point of this change is to avoid removing them and add a new handler which removes all sensitive data

Reply:
@harshavardhana the only way to avoid this, is by moving decryption code from GetObjectReader to GetObjectHandler again.. And by the way it makes sense for me, do you see any reason why shouldn't do that ?

/cc @donatello

Reply:
Well, I know why decryption is initialized inside GetObjectNInfo().

The reason is that GetObjectNInfo calls getObject() in a goroutine which needs an offset and a length to return data in encrypted format, but offset and length are actually calculated based on number of parts of that object, and parts information is actually retrieved in getObjectInfo().

Reply:
So let's resume the issue here,

GetObjectNInfo initializes reader with decryption enabled, during the initialization, SSE Key is removed from the headers, but then when we want to call rotateKeys() inside CopyObject handler, this latter won't find the key anymore.

Reply:
// cc @aead can explain why these headers are removed..

Reply:
> The rule is not clear when to remove and when not, so the point of this change is to avoid removing them and add a new handler which removes all sensitive data

This is correct. There is a lot of moving code regarding encryption. Centralizing this is on my TODO after becoming feature complete.

As a general rule-set:
- Confidential information **MUST** be removed as early as possible.
- Ideally there should be a central place for removing confidential information from requests.
- There must be filtering at the metadata level. Confidential information **MUST NEVER** be included in metadata.

On top encryption happens in handler-layer. object-layer has no clue whether it deals with encrypted data, that's not its job. (We must not mix layers)

BTW: A generic handler (`http.Handler`) is not the appropriated way to remove SSE headers. It would be too early since handler logic is executed before handler-layer (e.g. CopyObjectHandler).

Reply:
> BTW: A generic handler (`http.Handler`) is not the appropriated way to remove SSE headers. It would be too early since handler logic is executed before handler-layer (e.g. CopyObjectHandler).

Actually, the generic handler is executed before *and* after object handler layer, so here we remove SSE headers after the execution of the object handler layer

Reply:
> Actually, the generic handler is executed before and after object handler layer, so here we remove SSE headers after the execution of the object handler layer

Yes, but also *before* which causes the headers to be removed before passed to handler-layer, right?!

Reply:
> Yes, but also _before_ which causes the headers to be removed before passed to handler-layer, right?!

That was a mistake, I thought I successfly tested it but it seems it is not working.

Yes, before won't work, after won't also work.. because object-handlers already decide to send status code and headers and data.. it will be too late then to remove headers.

I did it differently now, it is in setCommonHeaders()